src/gatsby-node.ts

```diff
@@ -106,11 +107,37 @@ export const onPostBuild = ({ store }: any, userPluginOptions: PluginOptions) =>
     let routingRules: RoutingRule[] = [];
     let slsRoutingRules: ServerlessRoutingRule[] = [];
 
+    const temporaryRedirects = redirects.filter(redirect => redirect.fromPath !== '/')
+        .filter(redirect => !redirect.isPermanent)
+    
+    let permanentRedirects: GatsbyRedirect[] = redirects.filter(redirect => redirect.fromPath !== '/')
+        .filter(redirect => redirect.isPermanent)
+
+    if (pluginOptions.hostname && pluginOptions.protocol) {
+        const base = `${pluginOptions.protocol}://${pluginOptions.hostname}`
+        permanentRedirects = permanentRedirects.map((redirect) => {
+            return {
+                ...redirect,
+                toPath: resolve(base, redirect.toPath)
+            }
+        })
+    } else if(pluginOptions.hostname || pluginOptions.protocol) {
+        throw new Error(`Please either provide both 'hostname' and 'protocol', or neither of them.`)
+    }
+
     if (pluginOptions.generateRoutingRules) {
         routingRules = [
-            ...getRules(pluginOptions, redirects.filter(redirect => redirect.fromPath !== '/')),
+            ...getRules(pluginOptions, temporaryRedirects),
             ...getRules(pluginOptions, rewrites)
         ];
+        if (!pluginOptions.generateRedirectObjectsForPermanentRedirects) {
+            routingRules.push(...getRules(pluginOptions, permanentRedirects))
+            permanentRedirects = []
```

Comment:
I think this mutation of permanentRedirects makes the code harder to comprehend because the meaning of the variable is changed. Originally it means "all the permanent redirects" and now it's changed to mean "the permanent redirects that we want to create objects for".

I think a separate variable (`objectRedirects`) should be used for this. I also think instead of populating the variable by default and emptying it if `generateRedirectObjectsForPermanentRedirects` is false, we should create the array as empty by default and populate it if `generateRedirectObjectsForPermanentRedirects` is true. (I.e. in an else clause)